### ISSUE TITLE
Make queries deterministic when using read replicas

### ISSUE THREAD

**Issue by lutter** [graphprotocol/graph-node#4326]
When `graph-node` uses a shard that has asynchronous read replicas, query results are not deterministic because the read replicas can trail behind the main database, usually up to 30s.

When we execute a GraphQL query, we determine the block at which to execute by looking at the main database; fetching the data for the query result might be done against one of the read replicas which can lag behind the main database, and so even though we claim that the query was executed against block 100, the data that we use for the result might only be up to block 99.

**Comment by leoyvens** [comment]
@lutter Does this bug affect only queries by `latest`? Is the fix here for Graph Node to read the latest block from the same replica against which it will read the data?

**Comment by lutter** [comment]
Yes, that would be a way to fix that - we'd need to select the replica at the beginning of the query and use it to process everything. But that might be inconsistent with what we report as the latest block in the index-node API; to avoid reporting a latest block that you might not be able to get to, we'd have to take the min across all replicas and report that. Replication lag can add about 30s of delay though.

If we go with this fix, you could also observe the latest block going back if you put a `_meta { block { number } }` into your queries because different queries could go to different replicas.
